pkg/controller/controller_utils.go

```diff
@@ -706,19 +707,20 @@ func (s ByLogging) Less(i, j int) bool {
 		return len(s[i].Spec.NodeName) > 0
 	}
 	// 2. PodRunning < PodUnknown < PodPending
-	m := map[v1.PodPhase]int{v1.PodRunning: 0, v1.PodUnknown: 1, v1.PodPending: 2}
-	if m[s[i].Status.Phase] != m[s[j].Status.Phase] {
-		return m[s[i].Status.Phase] < m[s[j].Status.Phase]
+	if podPendingFirst[s[i].Status.Phase] != podPendingFirst[s[j].Status.Phase] {
+		return podPendingFirst[s[i].Status.Phase] > podPendingFirst[s[j].Status.Phase]
 	}
 	// 3. ready < not ready
 	if podutil.IsPodReady(s[i]) != podutil.IsPodReady(s[j]) {
 		return podutil.IsPodReady(s[i])
 	}
 	// TODO: take availability into account when we push minReadySeconds information from deployment into pods,
-	//       see https://github.com/kubernetes/kubernetes/issues/22065
+	//       see https://github.com/kubernetes/kubernetes/issues/4301
 	// 4. Been ready for more time < less time < empty time
-	if podutil.IsPodReady(s[i]) && podutil.IsPodReady(s[j]) && !podReadyTime(s[i]).Equal(podReadyTime(s[j])) {
-		return afterOrZero(podReadyTime(s[j]), podReadyTime(s[i]))
+	readyTime1 := podReadyTime(s[i])
+	readyTime2 := podReadyTime(s[j])
+	if podutil.IsPodReady(s[i]) && podutil.IsPodReady(s[j]) && !readyTime1.Equal(readyTime2) {
```

Comment:
I wonder what the proportion of calls to `Less` is where the condition `podutil.IsPodReady(s[i]) && podutil.IsPodReady(s[j]) ` is true.  If it is true in most calls, then this change decreases the number of times that `podReadyTime` is called.  If it is false in most calls, then this change _increases_ the number of times that `podReadyTime` is called, because logic short-circuiting would be preventing most of the calls to `podReadyTime` in the current code.

Reply:
Re-written related code so that the number of calls is always equal or fewer.